week7/cooper_client_react/04_saving_data.md

```diff
@@ -0,0 +1,252 @@
+
+```js
+// ...
+async saveCooperData() {
+    const result = this.calculate();
+    try {
+      await saveData(result);
+      this.props.entryHandler();
+    } catch(error) {
+      console.log(error);
+    }
+  }
+```
+
+The application will complain about `saveData()` not being defined. We need to add a new module to get this to work. Run:
+
+`$ touch src/Modules/PerformanceData.js`
+
+Add this:
+```js
+import axios from 'axios'
+import { storeAuthHeaders } from './Auth'
+
+const apiUrl = 'http://localhost:3000/api/v1';
+
+const saveData = (result) => {
+  const headers = JSON.parse(sessionStorage.getItem(['credentials']));
+  const path = apiUrl + '/performance_data';
+  return new Promise((resolve, reject) => {
+    axios.post(path, {
+      performance_data: { data: { message: result }}
+    }, {
+      headers: headers
+    })
+    .then(response => {
+      storeAuthHeaders(response);
+      resolve(response.data.message);
+    });  
+  });
+};
+
+export { saveData }
+```
+
+You have to import this file to the `DisplayCooperResult` component.
+
+`import { saveData } from '../Modules/PerformanceData';`
+
+So from our `DisplayCooperResult` component, we send in the result to this function. The first thing we do is grabbing the credentials we have stored in `sessionStorage`. Then we make a post request to the backend with those credentials and the cooper result. The response we get back gets sent back to saveCooperData function in the `DisplayCooperResult` component. We also call on the function `storeAuthHeaders` we have in the `Auth` module to update the credentials that we get in the response. If there is no error in the response, we call on the entryHandler. We have not defined the `entryHandler` yet, but let's do it now.
+
+Add this to `App` component:
+
+```js
+entryHandler() {
+  this.setState({ entrySaved: true });
+}
+```
+
+We also have to set that new state in our constructor:
+
+```js
+constructor(props) {
+    super(props);
+    this.state = {
+      // ...
+      entrySaved: false
+    }
+```
+
+Now we need to pass the entryHandler and `entrySaved` state to the `DisplayCooperResult` component:
+
+```js
+  <DisplayCooperResult
+    distance={this.state.distance}
+    gender={this.state.gender}
+    age={this.state.age}
+    authenticated={this.state.authenticated}
+    entrySaved={this.state.entrySaved}
+    entryHandler={this.entryHandler.bind(this)}
+  />
+```
+
+If you run the feature test now, you will get an error that says "Text not found "Your entry was saved". That's not weird, we haven't rendered the message we get from the backend anywhere. If you take a look at logs for the backend, you will see that the performance data is being saved correctly. We need to add and modify some of our code to get it to go green.
```

Comment:
Where can we see the logs for the backend?

Reply:
Done